src/databricks/labs/ucx/progress/history.py

```diff
@@ -0,0 +1,216 @@
+from __future__ import annotations
+import dataclasses
+import datetime as dt
+import json
+import logging
+from collections.abc import Iterable, Sequence
+from typing import ClassVar, Protocol, TypeVar, Generic, Any, get_type_hints
+
+from databricks.labs.lsql.backends import SqlBackend
+
+from databricks.labs.ucx.framework.owners import Ownership
+from databricks.labs.ucx.framework.utils import escape_sql_identifier
+from databricks.labs.ucx.progress.install import Historical
+
+logger = logging.getLogger(__name__)
+
+
+class DataclassWithIdAttributes(Protocol):
+    __dataclass_fields__: ClassVar[dict[str, Any]]
+
+    __id_attributes__: ClassVar[tuple[str, ...]]
+    """The names of attributes (can be dataclass fields or ordinary properties) that make up the object identifier.
+
+    All attributes must be (non-optional) strings.
+    """
+
+
+Record = TypeVar("Record", bound=DataclassWithIdAttributes)
+
+
+class HistoricalEncoder(Generic[Record]):
+    _job_run_id: int
```

Comment:
Why make these class attributes?

Reply:
This is one way to annotate the types of instance variables, although as a convenience we often do it during assignment from the initialiser (or even more commonly omit it entirely, leaving it to mypy to infer from the assigned value).

In this case I wanted to attach docstrings and this is the way to do that.

Reply:
sure, different to how we do it in the remainder code base, but fine